lib/broadway_cloud_pub_sub/client_acknowledger.ex

```diff
@@ -0,0 +1,284 @@
+  #     end
+
+  # These callbacks will be used by the ClientAcknowledger to dispatch requests
+  # to Google Cloud Pub/Sub.
+  @moduledoc false
+  alias Broadway.{Acknowledger, TermStorage}
+  alias BroadwayCloudPubSub.Client
+
+  @behaviour Acknowledger
+
+  @typedoc """
+  Ackmowledgement data for a `Broadway.Message`.
+  """
+  @type ack_data :: %{
+          :ack_id => String.t(),
+          optional(:on_failure) => ack_option,
+          optional(:on_success) => ack_option
+        }
+
+  @typedoc """
+  An acknowledgement action.
+  """
+  @type ack_option :: :ack | :ignore | :nack | {:nack, Client.ack_deadline()}
+
+  @type ack_ref :: reference
+
+  @type t :: %__MODULE__{
+          :client => module,
+          :client_opts => any,
+          :on_failure => ack_option,
+          :on_success => ack_option
+        }
+
+  @enforce_keys [:client]
+  defstruct [:client, :client_opts, on_failure: :ignore, on_success: :ack]
+
+  @doc """
+  Initializes this acknowledger for use with a `BroadwayCloudPubSub.Client`.
+
+  ## Options
+
+  The following options are usually provided by `BroadwayCloudPubSub.Producer`:
+
+    * `client` - The client module integrating with the #{inspect(__MODULE__)}.
+
+    * `on_success` - Optional. The action to perform for successful messages. Default is `:ack`.
+
+    * `on_failure` - The action to perform for failed messages. Default is `:ignore`.
+  """
+  @spec init(opts :: any) :: {:ok, t} | {:error, message :: binary}
+  def init(opts) do
+    with {:ok, client} <- validate(opts, :client),
+         {:ok, on_success} <- validate(opts, :on_success, :ack),
+         {:ok, on_failure} <- validate(opts, :on_failure, :ignore) do
+      {:ok,
+       %__MODULE__{
+         client: client,
+         on_failure: on_failure,
+         on_success: on_success
+       }}
+    end
+  end
+
+  @doc """
+  Creates a reference that can be passed to `acknowledger/2`.
+
+  The client configuration is stored in `Broadway.TermStorage`.
+  """
+  @spec ack_ref(config :: t, client_opts :: any) :: ack_ref
+  def ack_ref(config, client_opts) do
+    TermStorage.put(%{config | client_opts: client_opts})
+  end
+
+  @doc """
+  Returns an `acknowledger` to be put on a `Broadway.Message`.
+
+  ## Example
+
+      iex> BroadwayCloudPubSub.ClientAcknowledger.acknowledger("ackId", :ack_ref)
+      {BroadwayCloudPubSub.ClientAcknowledger, :ack_ref, %{ack_id: "ackId"}}
+
+  """
+  @spec acknowledger(ack_id :: Client.ack_id(), ack_ref) :: {__MODULE__, ack_ref, ack_data}
+  def acknowledger(ack_id, ack_ref) do
+    {__MODULE__, ack_ref, %{ack_id: ack_id}}
+  end
+
+  @impl Acknowledger
+  def ack(ack_ref, successful, failed) do
+    config = Broadway.TermStorage.get!(ack_ref)
+
+    success_actions = group_actions_ack_ids(successful, :on_success, config)
+    failure_actions = group_actions_ack_ids(failed, :on_failure, config)
+
+    success_actions
+    |> Map.merge(failure_actions, fn _, a, b -> a ++ b end)
+    |> ack_messages(config)
```

Comment:
Here the idea was to group the acknowledgements by action instead of by success/failure to reduce the number of API calls required.

Reply:
Perfect.